src/mainwin.cc

```diff
@@ -2013,6 +2017,166 @@ void MainWindow::csgRender()
+	if ( ! exportFile.open())
+	{
+		PRINT("Could not open temporary file.");
+		return;
+	}
+	QString exportFilename=exportFile.fileName();
+	
+	//Render the stl to a temporary file:
+	exportFileByName(this->root_geom, FileFormat::STL,
+		exportFilename.toLocal8Bit().constData(),
+		exportFilename.toUtf8());
+	
+	//Create a name that the order process will use to refer to the file. Base it off of the
+	//project name
+	QString userFacingName="unsaved.stl";
+	if (!this->fileName.isEmpty()) {
+		QString fileBasename = QFileInfo(this->fileName).baseName();
+		userFacingName = fileBasename.append("_").append(QString::number(printCounter++)).append(".stl");
+	}
+	
+	//Upload the file to the 3D Printing server and get the corresponding url to see it.
+	//The result is put in partUrl.
+	bool uploadWorked=uploadStlAndGetPartUrl(exportFilename, userFacingName, partUrl);
+	setCurrentOutput();
+
+	//Check the result:
+	if (! uploadWorked)
+	{
+		PRINT("An error occured while contacting the 3D Print API.");
+		return;
+	}
+	
+	//PRINT("partUrl:");
+	//PRINT(partUrl);
+	
+	//Open the url in the default browser:
+	QDesktopServices::openUrl ( partUrl );
+}
+
+//This function uploads an stl to the 3D printing API endpoint and returns a url that, 
+//when accessed, will show the stl file as a part that can be configured and added to the 
+//shopping cart.  Returns True if successful.
+//
+//Inputs:
+//    exportFilename  - The path to the temporary file that has the stl export in it.
+//    userFacingName  - Then name we should give the file when it is uploaded for the order process.
+//Outputs:
+//    partUrl         - The resulting url to go to next to continue the order process.
+bool MainWindow::uploadStlAndGetPartUrl(const QString & exportFilename, const QString &userFacingName, QUrl &partUrl)
+{
+	setCurrentOutput();
+	
+	//Create a request:
+	QNetworkRequest request(QUrl("https://print.openscad.org/api/v1/part-upload/"));
+	
+	//Set the content header:
+	request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
+	
+	//Get the file name from the file path:
+	QString fileNameBase=QFileInfo(exportFilename).fileName();
+	
+	//Create the request:
+	QJsonObject jsonInput;
+	
+	//Start building the json request:
+	jsonInput.insert("fileName", userFacingName);
+	
+	//Read our stl file:
+	QFile file(exportFilename);
+	if (!file.open(QIODevice::ReadOnly))
+	{
+		PRINT("Unable to open exported stl file.");
+		return false;
+	}
+	
+	//Convert it to base 64:
+	QString stlFileB64=file.readAll().toBase64();
+	
+	//Base 64-encoded file contents:
+	jsonInput.insert("file", stlFileB64);
+	
+	////Pring the input json:
+	//PRINT("Sending this JSON:");
+	//PRINT(QString(QJsonDocument(jsonInput).toJson()).toLocal8Bit().constData());
+	
+	//Create a network access manager:
+	QNetworkAccessManager nam;
+	
+	//Send the post:
+	QNetworkReply *reply = nam.post(request, QJsonDocument(jsonInput).toJson());
+
+	//Wait for the reply:
+	while(!reply->isFinished())
```

Comment:
Not in love with this. We should at least add a way to cancel it if the reply doesn't finish in reasonable time.

Cancellation could be offered through the progress bar (but that may be a bit confusing) or a dialog box which could show all the info that currently gets dumped to the console.

Reply:
Hmmm...  Yeah, I agree that we should have a better outcome if the API never responds.  I don't like the idea of poping up a dialog box or something that makes this feel like it's some bolted-on piece instead of part of the actual tool itself.  How about separating the code into lines that run before the api returns and lines that runs after the api returns, and hooking up some sort of callback to the finished signal?  http://doc.qt.io/qt-5/qnetworkaccessmanager.html#finished  I am not very adept with QT signals, but I assume they enable callbacks.  With that in place, at least, the rest of the software should still continue to work just as well while waiting for the reply.  If it never comes, no big deal.  If it comes 10 minutes later then the browser will pop up 10 minutes later, and they'll eventually figure out what happened.  Thoughts?  I'm up for editing things to make this better, I just want to agree on the direction before I dive in.

Reply:
One challenge with the progress bar is I don't think we have access to a % completion at this level.

Reply:
A background task could work as well. Just make sure that all the allocated memory gets dealt with. ..and make sure that nobody tries to access that temporary file after the QTemporaryFile object falls of the stack (as the file gets deleted).

I agree that a percent-wise progress bar doesn't make much sense. That's usually an option for existing progress bars..